src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java

```diff
@@ -314,6 +314,13 @@ public URLClassLoader run()
         try
         {
             checker.setClassLoader( projectClassLoader );
+            /*
+             * MCHECKSTYLE-387: If the previous method call was successful, emit a warning that the user is using
+             * an old version of checkstyle.
+             */
+            getLogger().warn( "Old version of checkstyle detected. Consider updating to >= v8.30" );
+            getLogger().warn( "For more information see: "
+                + "https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html" );
         }
         catch ( NoSuchMethodError ignored )
```

Comment:
CI show us that this ignore doesnot work and JVM is terminated.
Can we use reflection to check that method is actually present, and catch will not be required.

Reply:
@romani Can you post the error as you see it?  Based on a quick read here, I'm not sure how JVM would crash unless something is pointed out directly in checkstyle causing it.

@bmhm I think this is a bad way to check this.  This assumes just because projectClassLoader can be set that it's an old version.  I think it would make more logical sense to put a check early on in the plugin that checks the classpath version maven is aware of to indicate old version is being used.

Reply:
@hazendaz See https://github.com/checkstyle/checkstyle/issues/7190#issuecomment-592590991

Reply:
@rnveach Thanks I saw that but missed the important parts.  So really, the issue is not that the maven plugin doesn't handle properly at the lines this PR attempts to solve but rather subsequently on line 204 of the file.  A check needs added so this gracefully exists without failing at all.  I find this plugin execution optional in most ci/cd pipelines and it is horrible that it even fails hard like this.  The fix this PR does though is still not in the appropriate place.  Maven knows the plugins used, the versions used, etc.  It could check at start of the run and always warn that latest version it knows about is not being used.  It should then subsequently more gracefully handle failures.  The output here itself is good, but really just too far into the process when it could have been reported earlier.  In fact, it could even be set to skip run entirely if the version doesn't match the plugin and then actually start releasing this plugin to target checkstyle releases like it should be to start with.

Overall does this PR itself help, I guess, it will still fail so useful information will be present that helps general user but it just feels wrong and is allowing the problems to persist with this plugin in general instead of fixing the real issue.

Maybe a better conversation would be if maven should even be the owners of this plugin to start with.  Seems to me it might get steered better if it were under guidence of checkstyle team (proviced expertice with maven bits is there).  This is of course off topic a bit but worth expressing.

Reply:
>  I think this is a bad way to check this. 

The idea was not only to check the version, but rather to keep compatibility with old checkstyle versions which do need this method call.

> Maybe a better conversation would be if maven should even be the owners of this plugin to start with. 

The checkstyle (library) team said they do not have enough resources to file the PR themselves iirc. Also, the mentioned berthed was never deprecated because of this.

> Overall does this PR itself help,

I was only suggesting this because the library did not bother to keep any semantic versioning. Removing a method is over thing in the same major version, but also the checks from the checkstyle.xml file a user will configure differs. Users cannot just switch versions, because the library will just hard fail.

My previous commit tried to make the migration a little more graceful. I am sorry this did not work as intended. But this also means that all reviewers did not foresee a problem.

Reply:
@hazendaz 
I think @romani and I are talking about different things. Also, this issue is not the right place to discuss @romani s point. This PR is only to add a (maybe useless/needless) warning. But the try-catch-block is already in production. So we might want to move this discussion into a new issue?!

About the execution problem romani mentioned:
Please see https://github.com/checkstyle/checkstyle/issues/7190#issuecomment-593259487

When using the `maven-checkstyle-plugin:3.1.1`, you can execute both checkstyle(libary) versions up to 8.29 where the `setClassLoader()` method gets called, and after 8.29 (I tested 8.31-SNAPSHOT), java  will run into the `catch` block as expected.

The issue @romani is facing originates from not using a `maven-checkstyle-plugin` version with the try-catch-block, see https://github.com/checkstyle/checkstyle/issues/7190#issuecomment-593261410

Let's resolve this conversation with this. If you still encounter issues **after updating to `maven-checkstyle-plugin:3.1.1`**, please file a bug at apache jira: https://issues.apache.org/jira

@hazendaz Please open another conversation to talk about the idea of this PR. I am very open to not integrating it if there are better ways.

Reply:
My comment is false alarm, old version was used , see good comment at https://github.com/checkstyle/checkstyle/issues/7190#issuecomment-593261410 .

This review point is resolved. Thanks a lot for help. This commit change is good extra messages is always good for users.

Off topic but in reply to comments above:

>Seems to me it might get steered better if it were under guidence of checkstyle team (proviced expertice with maven bits is there). 

Yes, but there is a problem - checkstyle core library team is tiny. We have huge amount of PRs that are waiting for review each day. No matter how hard we work we never able to reach to zero active PRs for few years.  Each day we get new PRs. .... Maven plugin does not depends on metadata of core library that makes him skip huge amount of problems as other plugins has. All plugins that are eventually endup in our organization are plugins that heavily depends on checkstyle metadata about Checks, maintenance of them is hard ( I hope will be resolved ). 
It doesn't really matter where project is hosting, it is matter who maintain it. I can open gate and let somebody move project to be hosted in our organization, but if no extra heads and hands ... It will stay the same.
@bmhm , do you have more interest to contribute if it becomes non Apache project ?

>library did not bother to keep any semantic versioning. 

This is not true. We strictly(did bad version bump only ones) follow versioning except for first number ( it is reserved for huge update , like min jdk version bump). Second digit do mean update in API/metadata/... . Third digit means upgrade is safe for all plugins, only bugs we're fixed.

Reply:
Since I do not have that much time, I do not think I would contribute less or more than now (not contributing much anyway). Also, I just started contributing. Maybe you should ask the current maintainers instead (e.g. @elharo and @bdemers).

But I suggest you ask on the apache maven developer mailing list instead.

Reply:
This is how we proceed for years, a little from everyone, as sum there is some progress. :)